sdk/core/azure-core/inc/http/curl/curl.hpp

```diff
@@ -12,31 +12,199 @@
+     * line.
+     *
+     * @param buffer Points to a memory address with all or some part of a HTTP header.
+     * @param bufferSize Indicates the size of the buffer.
+     * @return Returns the index of the last parsed position from buffer. When the returned value is
+     * smaller than the body size, means there is part of the body response in the buffer.
+     */
+    size_t BuildHeader(uint8_t const* const buffer, size_t const bufferSize);
+
+  public:
+    /**
+     * @brief Construct a new Response Buffer Parser object.
+     * Set the initial state and parsing completion.
+     *
+     */
+    ResponseBufferParser()
+    {
+      state = ResponseParserState::StatusLine;
+      parseCompleted = false;
+    }
+
+    // Parse contents of buffer to construct HttpResponse. Returns the index of the last parsed
+    // possition. Return bufferSize when all buffer was used to parse
+    /**
+     * @brief Parses the content of a buffer to constuct a valid HTTP Response. This method is
+     * expected to be called over and over until it returns 0, indicating there is nothing more to
+     * parse to build the HTTP Response.
+     *
+     * @param buffer points to a memory area that contains, all or some part of an HTTP response.
+     * @param bufferSize Indicates the size of the buffer.
+     * @return Returns the index of the last parsed position. Returning a 0 means nothing was parsed
+     * and it is likely that the HTTP Response is completed. Returning the same value as the buffer
+     * size means all buffer was parsed and the HTTP might be completed or not. Returning a value
+     * smaller than the buffer size will likely indicate that the HTTP Response is completed and
+     * that the rest of the buffer contains part of the response body.
+     */
+    size_t Parse(uint8_t const* const buffer, size_t const bufferSize);
+
+    /**
+     * @brief Indicates when the parser has completed parsing and building the HTTP Response.
+     *
+     * @return true if parsing is completed. Otherwise false.
+     */
+    bool IsParseCompleted() const { return this->parseCompleted; }
 
-    // for every client instance, create a default response
-    std::unique_ptr<Azure::Core::Http::Response> m_response;
-    bool m_firstHeader;
+    /**
+     * @brief Moves the internal response to a different owner.
+     *
+     * @return Will move the response only if parsing is completed and if the HTTP Response was not
+     * moved before.
+     */
+    std::unique_ptr<Response> GetResponse()
+    {
+      if (this->parseCompleted && this->m_response != nullptr)
+      {
+        return std::move(this->m_response);
+      }
+      return nullptr; // parse is not completed or response has been moved already.
+    }
+  };
 
+  class CurlSession {
+  private:
     CURL* m_pCurl;
+    curl_socket_t m_curlSocket; // For Stream implementartion
+
+    std::unique_ptr<Response> m_response;
+    Request& m_request;
```

Comment:
Ref type members are not allowed
https://azure.github.io/azure-sdk/cpp_design.html#const-and-reference-members